pygmt/datatypes.py

```diff
@@ -0,0 +1,101 @@
+        ("z_units", ctp.c_char * GMT_GRID_UNIT_LEN80),  # grid value units
+        ("title", ctp.c_char * GMT_GRID_TITLE_LEN80),  # name of data set
+        ("command", ctp.c_char * GMT_GRID_COMMAND_LEN320),  # name of generating command
+        ("remark", ctp.c_char * GMT_GRID_REMARK_LEN160),  # comments for this data set
+        # items used internally by GMT
+        # number of data points (n_columns * n_rows) [padding is excluded]
+        ("nm", ctp.c_size_t),
+        # actual number of items (not bytes) required to hold this grid (mx * my)
+        ("size", ctp.c_size_t),
+        ("bits", ctp.c_uint),  # bits per data value
+        # complex grid
+        # 0 for normal
+        # GMT_GRID_IS_COMPLEX_REAL = real part of complex grid
+        # GMT_GRID_IS_COMPLEX_IMAG = imag part of complex grid
+        ("complex_mode", ctp.c_uint),
+        ("type", ctp.c_uint),  # grid format
+        ("n_bands", ctp.c_uint),  # number of bands [1]
+        ("mx", ctp.c_uint),  # actual x-dimension in memory, allowing for padding
+        ("my", ctp.c_uint),  # actual y-dimension in memory, allowing for padding
+        ("pad", ctp.c_uint * 4),  # padding on west, east, south, north sides [2,2,2,2]
+        # Three or four char codes T|B R|C S|R|S (grd) or B|L|P + A|a (img)
+        # describing array layout in mem and interleaving
+        ("mem_layout", ctp.c_char * 4),
+        ("nan_value", gmt_grdfloat),  # missing value as stored in grid file
+        ("xy_off", ctp.c_double),  # 0.0 for gridline grid and 0.5 for pixel grid
+        ("ProjRefPROJ4", ctp.c_char_p),  # referencing system string in PROJ.4 format
+        ("ProjRefWKT", ctp.c_char_p),  # referencing system string in WKT format
+        ("ProjRefEPSG", ctp.c_int),  # referencing system EPSG code
+        ("hidden", ctp.c_void_p),  # lower-level information for GMT use only
+    ]
+
+
+class GMT_GRID(ctp.Structure):
+    """
+    Structure for GMT grid.
+    """
+
+    _fields_ = [
+        ("header", ctp.POINTER(GMT_GRID_HEADER)),  # pointer to grid header
+        ("data", ctp.POINTER(gmt_grdfloat)),  # pointer to grid data
+        ("x", ctp.POINTER(ctp.c_double)),  # pointer to x coordinate vector
+        ("y", ctp.POINTER(ctp.c_double)),  # pointer to y coordinate vector
+        ("hidden", ctp.c_void_p),  # low-level information for GMT use only
+    ]
+
+    def to_dataarray(self):
+        header = self.header.contents
+        pad = header.pad[:]
+        x = self.x[: header.n_columns]
+        y = self.y[: header.n_rows]
+        data = np.array(self.data[: header.mx * header.my]).reshape(
+            (header.my, header.mx)
+        )[pad[2] : header.my - pad[3], pad[0] : header.mx - pad[1]]
+        if x[0] > x[1]:
+            x = list(reversed(x))
+            data = np.fliplr(data)
+        if y[0] > y[1]:
+            y = list(reversed(y))
+            data = np.flipud(data)
+        return xr.DataArray(data, coords=[y, x], dims=["lat", "lon"])
```

Comment:
The GMT_GRID object has the dimension names in `header.x_units` and `header.y_units` (e.g., `longitude [degrees_east]` and `latitude [degrees_north]`. So we should be able to specify the dimension names based on these values.

I just did a few tests below:

1. Reading a geographic grid using xarray gives:
```python
In [1]: import xarray as xr

In [2]: grid = xr.open_dataarray("earth_relief_01d_g.grd")

In [3]: grid
Out[3]: 
<xarray.DataArray 'z' (lat: 181, lon: 361)>
[65341 values with dtype=float32]
Coordinates:
  * lon      (lon) float64 -180.0 -179.0 -178.0 -177.0 ... 178.0 179.0 180.0
  * lat      (lat) float64 -90.0 -89.0 -88.0 -87.0 -86.0 ... 87.0 88.0 89.0 90.0
Attributes:
    actual_range:  [-8600.5  5559. ]
    long_name:     elevation (m)
```

2. I also tried to generate a Cartesian grid with `xyz2grd`:
```
gmt xyz2grd -R0/1/0/1 -I1 -Gcart.nc << EOF
0 0 0
0 1 1
1 0 2
1 1 3
EOF
```
Running `gmt grdinfo cart.nc` gives:
```
cart.nc: Title: 
cart.nc: Command: gmt xyz2grd -R0/1/0/1 -I1 -Gcart.nc
cart.nc: Remark: 
cart.nc: Gridline node registration used [Cartesian grid]
cart.nc: Grid file format: nf = GMT netCDF format (32-bit float), CF-1.7
cart.nc: x_min: 0 x_max: 1 x_inc: 1 name: x n_columns: 2
cart.nc: y_min: 0 y_max: 1 y_inc: 1 name: y n_rows: 2
cart.nc: v_min: 0 v_max: 3 name: z
cart.nc: scale_factor: 1 add_offset: 0
cart.nc: format: classic
cart.nc: Default CPT:
```
Loading the grid using xarray gives:
```
<xarray.DataArray 'z' (y: 2, x: 2)>
[4 values with dtype=float32]
Coordinates:
  * x        (x) float64 0.0 1.0
  * y        (y) float64 0.0 1.0
Attributes:
    long_name:     z
    actual_range:  [0. 3.]
```

3. Also tried to change the x- and y-names to other values:
```
$ gmt xyz2grd -R0/1/0/1 -I1 -D+xnamex+ynamey -Gcart.nc << EOF
0 0 0
0 1 1
1 0 2
1 1 3
EOF
$ gmt grdinfo cart.nc 
cart.nc: Title: 
cart.nc: Command: gmt xyz2grd -R0/1/0/1 -I1 -D+xnamex+ynamey -Gcart.nc
cart.nc: Remark: 
cart.nc: Gridline node registration used [Cartesian grid]
cart.nc: Grid file format: nf = GMT netCDF format (32-bit float), CF-1.7
cart.nc: x_min: 0 x_max: 1 x_inc: 1 name: namex n_columns: 2
cart.nc: y_min: 0 y_max: 1 y_inc: 1 name: namey n_rows: 2
cart.nc: v_min: 0 v_max: 3 name: z
cart.nc: scale_factor: 1 add_offset: 0
cart.nc: format: classic
cart.nc: Default CPT: 
```
Loading the grid with xarray gives:
```
<xarray.DataArray 'z' (y: 2, x: 2)>
[4 values with dtype=float32]
Coordinates:
  * x        (x) float64 0.0 1.0
  * y        (y) float64 0.0 1.0
Attributes:
    long_name:     z
    actual_range:  [0. 3.]
```

My guess is, xarray always maps dimension names "longitude/latitude" to "lon"/"lat" and other names to "x"/"y". And there may be some conventions (e.g., https://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#coards-relationship)

Reply:
What function are you using to "load the grid with xarray"? Is it [`xr.open_dataarray`](https://docs.xarray.dev/en/v2023.04.2/generated/xarray.open_dataarray.html) or [`pygmt.load_dataarray`](https://www.pygmt.org/v0.9.0/api/generated/pygmt.load_dataarray.html)? The .nc grid from xyz2grd doesn't have dimension names, so xarray would fallback to `x`/`y`, but if you read a NetCDF file that has properly set the dimension names (e.g. those at https://github.com/pydata/xarray-data), then you should get `lat`/`lon` or some other names instead.

Anyways, I'm ok with hardcoding the dimension names to `x`/`y` for now since that's what the current behaviour is, and we can work on preserving the dimension names (i.e. fix #1819) later.